src/Bicep.Core/Semantics/SemanticModel.cs

```diff
@@ -39,29 +40,26 @@ public class SemanticModel : ISemanticModel
         private readonly Lazy<ImmutableArray<DeclaredResourceMetadata>> declaredResourcesLazy;
         private readonly Lazy<ImmutableArray<IDiagnostic>> allDiagnostics;
 
-        public SemanticModel(Compilation compilation, BicepSourceFile sourceFile, IFileResolver fileResolver, IBicepAnalyzer linterAnalyzer, RootConfiguration configuration, IFeatureProvider features, IApiVersionProvider apiVersionProvider)
+        public SemanticModel(Compilation compilation, BicepSourceFile sourceFile, IFileResolver fileResolver, IBicepAnalyzer linterAnalyzer, RootConfiguration configuration, IFeatureProvider features)
         {
             Trace.WriteLine($"Building semantic model for {sourceFile.FileUri} ({sourceFile.FileKind})");
 
-            Compilation = compilation;
-            SourceFile = sourceFile;
-            Configuration = configuration;
-            Features = features;
-            ApiVersionProvider = apiVersionProvider;
-            FileResolver = fileResolver;
+            this.Compilation = compilation;
+            this.SourceFile = sourceFile;
+            this.Configuration = configuration;
+            this.Features = features;
+            this.FileResolver = fileResolver;
 
             // create this in locked mode by default
             // this blocks accidental type or binding queries until binding is done
             // (if a type check is done too early, unbound symbol references would cause incorrect type check results)
-            var symbolContext = new SymbolContext(compilation, this);
-            SymbolContext = symbolContext;
+            this.SymbolContext = new SymbolContext(compilation, this);
+            this.Binder = new Binder(compilation.NamespaceProvider, features, sourceFile, this.SymbolContext);
+            this.ApiVersionProvider = new ApiVersionProvider(features, this.Binder.NamespaceResolver.GetAvailableResourceTypes());
+            this.TypeManager = new TypeManager(features, this.Binder, fileResolver, this.ParsingErrorLookup, this.SourceFile.FileKind);
 
-            Binder = new Binder(compilation.NamespaceProvider, features, sourceFile, symbolContext);
-            TypeManager = new TypeManager(features, Binder, fileResolver, this.ParsingErrorLookup, this.SourceFile.FileKind);
-
-            // name binding is done
-            // allow type queries now
-            symbolContext.Unlock();
+            // name binding is done allow type queries now
+            ((SymbolContext)this.SymbolContext).Unlock();
```

Comment:
nit: not a fan of this type of assumption for type.  IMHO better to keep a local variable copy of SymbolContext typed SymbolContext and use it here.

Reply:
I don't have a strong opinion, this code was for the most part untouched and has no behavior change from when it was originally submitted except for the comment thay I adressed from @shenglol
I am happy to change it but I don't understand what you mean by "keep a local variable copy of SymbolContext typed SymbolContext and use it here", happy to do the change though... :)